### PR TITLE
Allow `utils.derive_from` to accept functions, apply across array

### LINKED ISSUE #4758: Use derived_from logic in dask array
Dask Dataframe methods copy Pandas docstrings, but with some logic to add disclaimers around unsupported arguments (and other niceties).  It would be useful to reuse this logic in dask array, which often uses functools.wraps.  I ran into this in dask.array.routines, but I suspect that it is in lots of places.

The `derived_from` code is currently baked around class methods, which isn't always appropriate for dask array (many of which are top level functions), but probably a lot of the lower level bits would work fine.  We should...

1.  Rearrange the logic in derived_from to also support top level functions, making a wraps-like decorator function
2.  Replace functools.wraps with that decorator

### PR DESCRIPTION
**Pull request by martindurant**
To be applied to numpy top-level functions, as seen throughout the dask.array
package. One test case so far: da.routines.array .

- [x] Tests added / passed
- [x] Passes `flake8 dask`

Fixes #4758

### CODE REVIEW

File: dask/array/routines.py [hunk 1]
```diff
@@ -1202,7 +1203,7 @@ def piecewise(x, condlist, funclist, *args, **kw):
     )
 
 
-@wraps(chunk.coarsen)
+@wraps(chunk.coarsen)  # this is not a numpy function
```
Comment (mrocklin) [inline dask/array/routines.py]: This comment may not make as much sense outside of th context of this PR

Comment (martindurant) [reply dask/array/routines.py]: I meant to write a comment into the PR, that there are a couple of annotations in the diff, meaning they can be discussed here and removed when resolved. I guess I've done that now :)

**Comment by jrbourbeau** [comment]
Hmm I'm not able to locally reproduce the error in the Python 2.7 Travis build. Are others able to get the raised `AssertionError`?

**Comment by martindurant** [comment]
It's due to running with optimisation, which strips the original docstrings. Ignoring py2 for these tests in the rest of this PR.

**Comment by jrbourbeau** [comment]
> It's due to running with optimisation, which strips the original docstrings

Good to know! I had no idea about this behavior

**Comment by mrocklin** [comment]
In general the code solution here looks great to me.  I'm happy with how simple this ended up being.  I spot checked this in IPython and ran into a small issue.  It looks like spacing is off.  I think that IPython dedents everything if spacing is consistent.

### Master

```
Signature: da.tensordot(a, b, axes=2)
Docstring:
Compute tensor dot product along specified axes for arrays >= 1-D.

Given two tensors (arrays of dimension greater than or equal to one),
`a` and `b`, and an array_like object containing two array_like
...
```

### This branch
```
Signature: da.tensordot(lhs, rhs, axes=2)
Docstring:
    Compute tensor dot product along specified axes for arrays >= 1-D.

    This docstring was copied from numpy.tensordot.

    Some inconsistencies with the Dask version may exist.

    Given two tensors (arrays of dimension greater than or equal to one),
    `a` and `b`, and an array_like object containing two array_like
    ...
```

It seems like this is due to some code lines not being indented to the same amount

```
    >>> c  # doctest: +SKIP
    array([[ 4400.,  4730.],
           [ 4532.,  4874.],
           [ 4664.,  5018.],
           [ 4796.,  5162.],
           [ 4928.,  5306.]])
>>> # A slower but equivalent way of computing the same...
    >>> d = np.zeros((5,2))  # doctest: +SKIP
    >>> for i in range(5):  # doctest: +SKIP
    ...   for j in range(2):
```

File: dask/utils.py [hunk 2]
```diff
@@ -441,7 +441,7 @@ def _skip_doctest(line):
     # NumPy docstring contains cursor and comment only example
     stripped = line.strip()
     if stripped == '>>>' or stripped.startswith('>>> #'):
-        return stripped
+        return line
```
Comment (martindurant) [inline dask/utils.py]: @mrocklin , any idea why this was so? The example in `np.tensordot` may have been the only one with `">>> #"` in the docstring...

**Comment by TomAugspurger** [comment]
FWIW, I don't see the indentation issue

```
>>> c  # doctest: +SKIP
array([[ 4400.,  4730.],
       [ 4532.,  4874.],
       [ 4664.,  5018.],
       [ 4796.,  5162.],
       [ 4928.,  5306.]])
>>> # A slower but equivalent way of computing the same...
>>> d = np.zeros((5,2))  # doctest: +SKIP
>>> for i in range(5):  # doctest: +SKIP
...   for j in range(2):
...     for k in range(3):
...       for n in range(4):
...         d[i,j] += a[k,n,i] * b[n,k,j]
>>> c == d  # doctest: +SKIP
array([[ True,  True],
       [ True,  True],
       [ True,  True],
       [ True,  True],
       [ True,  True]])


```

```
In [7]: IPython.__version__
Out[7]: '7.5.0'

In [8]: np.__version__
Out[8]: '1.16.2'
```

File: dask/utils.py (hunk 2 above)
Comment (TomAugspurger) [reply dask/utils.py]: Goes back to https://github.com/dask/dask/pull/1669/files#diff-64b17ee9155433cad463c80a971a3223L547. I think if the CI is passing then we're OK with this change.

**Pull request merged**